File: 2766-find-the-prefix-common-array-of-two-arrays/2766-find-the-prefix-common-array-of-two-arrays.dart
class Solution {
  List<int> findThePrefixCommonArray(List<int> A, List<int> B) {
    int n = A.length;
    List<int> ans = List.filled(n, 0);
    List<bool> presentA = List.filled(n + 1, false);
    List<bool> presentB = List.filled(n + 1, false);
    int cnt = 0;

    for (int i = 0; i < n; i++) {
      if (A[i] == B[i]) {
        cnt++;
      }
      if (presentA[B[i]]) {
        cnt++;
      }
      if (presentB[A[i]]) {
        cnt++;
      }
      presentA[A[i]] = true;
      presentB[B[i]] = true;
      ans[i] = cnt;
    }

    return ans;
  }
}

File: 2766-find-the-prefix-common-array-of-two-arrays/2766-find-the-prefix-common-array-of-two-arrays.java
class Solution {
    public int[] findThePrefixCommonArray(int[] A, int[] B) {
        int n = A.length;
        int[] ans = new int[n];
        boolean[] pa = new boolean[n + 1];  
        boolean[] pb = new boolean[n + 1];
        int cnt = 0;

        for (int i = 0; i < n; i++) {
            cnt += A[i] == B[i] ? 1 : 0;
            cnt += pa[B[i]] ? 1 : 0;
            cnt += pb[A[i]] ? 1 : 0;

            pa[A[i]] = true;
            pb[B[i]] = true;
            ans[i] = cnt;
        }
        return ans;
    }
}